executor/api/rest/server.go

```diff
@@ -344,3 +346,30 @@ func (r *SeldonRestApi) predictions(w http.ResponseWriter, req *http.Request) {
 	}
 	r.respondWithSuccess(w, http.StatusOK, resPayload)
 }
+
+func (r *SeldonRestApi) graph_metadata(w http.ResponseWriter, req *http.Request) {
+	r.Log.Info("Graph Metadata called.")
+
+	ctx := req.Context()
+
+	// Apply tracing if active
+	if opentracing.IsGlobalTracerRegistered() {
```

Comment:
Do we need / want tracing on metadata? Has this code path been tested? If not it may be best to remove it for now (or make sure we test it with jaeger)

Reply:
I am happy to remove it - just when tried initially to do without this block I was getting some errors so I basically copied it as the boiler plate from other endpoints.

Reply:
All other endpoints seems to be having the tracing and they don't seem to have any special tests for it. If I do remove this block I get bunch of ugly errors:
```
{"level":"info","ts":1589987540.2213755,"logger":"SeldonRestApi","msg":"Graph Metadata called."}
{"level":"info","ts":1589987540.221461,"logger":"JSONRestClient","msg":"Calling HTTP","URL":"http://0.0.0.0:9001/metadata"}
2020/05/20 16:12:20 http: panic serving 127.0.0.1:42208: runtime error: invalid memory address or nil pointer dereference
goroutine 40 [running]:
net/http.(*conn).serve.func1(0xc0003c4000)
	/usr/lib/go-1.14/src/net/http/server.go:1772 +0x139
panic(0x1580900, 0x25f3140)
	/usr/lib/go-1.14/src/runtime/panic.go:975 +0x3e3
github.com/seldonio/seldon-core/executor/api/rest.(*JSONRestClient).doHttp(0xc000143720, 0x19d7560, 0xc0003c2240, 0xc00036c3f0, 0x7, 0x17879e9, 0x9, 0xc0000b2280, 0x0, 0x0, ...)
	/home/rskolasinski/work/seldon-core/executor/api/rest/client.go:168 +0x830
github.com/seldonio/seldon-core/executor/api/rest.(*JSONRestClient).call(0xc000143720, 0x19d7560, 0xc0003c2240, 0xc00036c3f0, 0x7, 0x17879e9, 0x9, 0xc00036c3e0, 0x7, 0x2329, ...)
	/home/rskolasinski/work/seldon-core/executor/api/rest/client.go:228 +0x251
github.com/seldonio/seldon-core/executor/api/rest.(*JSONRestClient).Metadata(0xc000143720, 0x19d7560, 0xc0003c2240, 0xc00036c3f0, 0x7, 0xc00036c3e0, 0x7, 0x2329, 0x0, 0x0, ...)
	/home/rskolasinski/work/seldon-core/executor/api/rest/client.go:238 +0x1c7
github.com/seldonio/seldon-core/executor/predictor.(*PredictorProcess).MetadataMap(0xc0004656a0, 0xc00014ca00, 0xc000314000, 0xc000314000, 0x0)
	/home/rskolasinski/work/seldon-core/executor/predictor/predictor_process.go:335 +0xa6
github.com/seldonio/seldon-core/executor/predictor.NewGraphMetadata(0xc00045b6a0, 0xc0000ee120, 0xa, 0x19e0ec0, 0xc0003e4060)
	/home/rskolasinski/work/seldon-core/executor/predictor/metadata.go:30 +0x3d
github.com/seldonio/seldon-core/executor/api/rest.(*SeldonRestApi).graph_metadata(0xc0000ee360, 0x7fb25c62d380, 0xc0007b4050, 0xc0003cc200)
	/home/rskolasinski/work/seldon-core/executor/api/rest/server.go:364 +0x219
net/http.HandlerFunc.ServeHTTP(0xc000372730, 0x7fb25c62d380, 0xc0007b4050, 0xc0003cc200)
	/usr/lib/go-1.14/src/net/http/server.go:2012 +0x44
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerDuration.func1(0x19d4520, 0xc0003d6000, 0xc0003cc200)
	/home/rskolasinski/go/pkg/mod/github.com/prometheus/client_golang@v1.3.0/prometheus/promhttp/instrument_server.go:68 +0x11c
net/http.HandlerFunc.ServeHTTP(0xc000479d10, 0x19d4520, 0xc0003d6000, 0xc0003cc200)
	/usr/lib/go-1.14/src/net/http/server.go:2012 +0x44
github.com/seldonio/seldon-core/executor/api/rest.(*CloudeventHeaderMiddleware).Middleware.func1(0x19d4520, 0xc0003d6000, 0xc0003cc200)
	/home/rskolasinski/work/seldon-core/executor/api/rest/server.go:183 +0xd4
net/http.HandlerFunc.ServeHTTP(0xc0003e4000, 0x19d4520, 0xc0003d6000, 0xc0003cc200)
	/usr/lib/go-1.14/src/net/http/server.go:2012 +0x44
github.com/seldonio/seldon-core/executor/api/rest.puidHeader.func1(0x19d4520, 0xc0003d6000, 0xc0003cc200)
	/home/rskolasinski/work/seldon-core/executor/api/rest/server.go:198 +0x111
net/http.HandlerFunc.ServeHTTP(0xc0003e4020, 0x19d4520, 0xc0003d6000, 0xc0003cc200)
	/usr/lib/go-1.14/src/net/http/server.go:2012 +0x44
github.com/gorilla/mux.(*Router).ServeHTTP(0xc0001fc000, 0x19d4520, 0xc0003d6000, 0xc0003cc000)
	/home/rskolasinski/go/pkg/mod/github.com/gorilla/mux@v1.7.3/mux.go:212 +0xe2
net/http.serverHandler.ServeHTTP(0xc0001f42a0, 0x19d4520, 0xc0003d6000, 0xc0003cc000)
	/usr/lib/go-1.14/src/net/http/server.go:2807 +0xa3
net/http.(*conn).serve(0xc0003c4000, 0x19d74a0, 0xc0003c8000)
	/usr/lib/go-1.14/src/net/http/server.go:1895 +0x86c
created by net/http.(*Server).Serve
	/usr/lib/go-1.14/src/net/http/server.go:2933 +0x35c
```


Same if I remove that block for let say predictions endpoint. It starts to fail then on requesting predictions but tests passes. It seems this is an independent problem which we should address in separate issue maybe/

Reply:
It seems that client code expect tracing to be on/off for all requests. It is failing on `opentracing.StartSpan` in 
```golang
	if opentracing.IsGlobalTracerRegistered() {
		tracer := opentracing.GlobalTracer()

		parentSpan := opentracing.SpanFromContext(ctx)
		clientSpan := opentracing.StartSpan(
			method,
			opentracing.ChildOf(parentSpan.Context()))
		defer clientSpan.Finish()
		tracer.Inject(clientSpan.Context(), opentracing.HTTPHeaders, opentracing.HTTPHeadersCarrier(req.Header))
	}
```

Reply:
Oh, what I meant by testing, is mainly whether you had actually run it so that this code path is run, ie running a model with jaeger enabled. Un regards to it failing if it's removed, that's strange! I totall agree if it's a bigger issue we should move it to a different issue, but definitely one to just confirm. If we leave this code it may be worth just trying to deploy a model with tracing enabled (and jaeger running) which should tell us if it works or not

Reply:
It seems that for me `opentracing.IsGlobalTracerRegistered()` always evaluates to `true` even if I just run executor as 
```
go run ../../../main.go --sdep seldon-model --namespace default --predictor example --file ./chain.yaml --http_port 8000
```

Reply:
I does to seem work fine though:
![image](https://user-images.githubusercontent.com/10928117/82494344-53a71780-9ae1-11ea-826d-012fb07851de.png)

Reply:
https://github.com/SeldonIO/seldon-core/issues/1858